### ISSUE TITLE
Undefined variable in tutorial

### ISSUE THREAD

**Issue by AuracleTech** [KyleMayes/vulkanalia#143]
[Here](https://kylemayes.github.io/vulkanalia/setup/validation_layers.html#using-validation-layers) you provide this code
```rs
let info = vk::InstanceCreateInfo::builder()
    .application_info(&application_info)
    .enabled_layer_names(&layers)
    .enabled_extension_names(&extensions)
    .flags(flags);
 ```
 But the variable `flags` from `.flags(flags);` is undefined

**Comment by AuracleTech** [comment]
[Here](https://kylemayes.github.io/vulkanalia/presentation/window_surface.html)
`vk_window::create_surface` doesn't exist in favor of `vulkanalia::window::create_surface;`
The new function takes 3 arguments one of which is `display` which is missing in the tutorial

**Comment by AuracleTech** [comment]
[Here](https://kylemayes.github.io/vulkanalia/presentation/swapchain.html)
`Ok(extensions)` uses an unknown variable, it should be `Ok(())`

**Comment by KyleMayes** [comment]
For the first problem, I checked this and `flags` is defined in the previous chapter (`Instance`).

The other two problems are duplicates of existing issues that I've just now resolved with some updates to the tutorial.

Thank you for the reports!

**Comment by AuracleTech** [comment]
Oh I didn't do this part because it said `Driver compability (macOS)`
I thought it was a part only for macOS users

**Comment by KyleMayes** [comment]
That's a good point, I forgot that the section that introduces the `flags` variable is marked as being a `macOS` section.

That could definitely be improved, I see why someone would miss it now.
I'll think about how to update that section to improve it.

**Comment by AuracleTech** [comment]
[Here](https://kylemayes.github.io/vulkanalia/pipeline/fixed_functions.html) is says this :

And then create the object in the `create_pipeline` function:
```rs
let layout_info = vk::PipelineLayoutCreateInfo::builder();
data.pipeline_layout = device.create_pipeline_layout(&layout_info, None)?;
```
The `create_pipeline` at that point was pretty populated so I placed it near the end before `device.destroy_shader_module`
I did not know I had to instantiate the layout before the pipeline was created because the pipeline use `data.pipeline_layout`
This caused me an error

Maybe it could precise to add the function at the top of `create_pipeline` or even directly in App create before `create_pipeline`

**Comment by AuracleTech** [comment]
[Here](https://kylemayes.github.io/vulkanalia/drawing/command_buffers.html) It mentions `None of these flags are applicable for us right now.`
but *I believe* we do need `vk::CommandBufferUsageFlags::SIMULTANEOUS_USE` because I was getting spammed with errors without it

In `create_command_buffers` I changed
```rs
.flags(vk::CommandBufferUsageFlags::empty()) // Optional.
```
to
```rs
.flags(vk::CommandBufferUsageFlags::SIMULTANEOUS_USE) // Optional.
```

Hopefully this helps and again sorry for lots all the report you tell me if you would prefer I open a different issue for each or whatever you'd prefer

**Comment by KyleMayes** [comment]
> Hopefully this helps and again sorry for lots all the report you tell me if you would prefer I open a different issue for each or whatever you'd prefer

Not a problem at all, thanks for the reports. I'm sure you aren't the first person to have run into these issues so I'm glad to know about them (its been quite a while since I went through each chapter and checked that it all hangs together, compiles, etc.).

A single issue is fine, I'll hope to start fixing/clarifying these problems soon.

**Comment by AuracleTech** [comment]
[Here](https://kylemayes.github.io/vulkanalia/vertex/vertex_buffer_creation.html#buffer-creation) it says
 > The flags parameter is used to configure sparse buffer memory, which is not relevant right now. We'll leave it at the default value of 0.

Issue is there's no flag parameter, I think this is the equivalent of dead code but for docs

**Comment by AuracleTech** [comment]
[Here](https://kylemayes.github.io/vulkanalia/uniform/descriptor_pool_and_sets.html#using-descriptor-sets) is missing a critical change for `layout_info`

This line
```rs
let layout_info = vk::PipelineLayoutCreateInfo::builder();
```
Requires a change to add the newly added `descriptor_set_layout`
```rs
let binding = &[data.descriptor_set_layout];
let layout_info = vk::PipelineLayoutCreateInfo::builder().set_layouts(binding);
```

**Comment by AuracleTech** [comment]
[Here](https://kylemayes.github.io/vulkanalia/swapchain/recreation.html) is also missing a critical line
```rs
if changed {
    self.recreate_swapchain(window)?;
    self.resized = false; // MISSING LINE
} else if let Err(e) = result {
    return Err(anyhow!(e));
}
```
I was wondering why I had 120 fps on average even though my monitor is 280hz.
The swapchain was getting recreated every frame 💀 Im now back to 1000+ fps

**Comment by KyleMayes** [comment]
Creating a checklist so I can track this more easily:

- [ ] Update instance chapter to make it clear the driver compatibility section applies to everyone, not just macOS
- [ ] Update fixed functions chapter to make placement of pipeline layout initialization more clear
- [ ] Investigate whether simultaneous use command buffer flag should be included in command buffers chapter
- [ ] Remove reference to unused sparse buffer memory flag in vertex buffer creation chapter
- [ ] Fix descriptor set layout initialization in descriptor pool chapter
- [ ] Fix reference to sample count flags in texture images chapter
- [ ] Fix resizing logic in swapchain recreation chapter
- [ ] Fix `copy_bufer_to_image` typo in texture images chapter

Will start working on these this weekend, thank you again for documenting these.

**Comment by AuracleTech** [comment]
It's a recommendation only but I recommend changing from `glm` to `cgmath`
Doing so might require lots of editing ( somewhat ) of the tutorial but could improve the quality of it
`glm` hasn't been updated for 6 years and due to it I was was getting lots of security issues from `Dependabot`

![glm dependencies issues](https://user-images.githubusercontent.com/40531871/232262238-92d69f47-7154-483c-8d3e-cb4f9a51c535.png)

This is the equivalent of `glm` but using `cgmath`

```rs
// Vertex struct
pub(crate) pos: cgmath::Vector3<f32>,
pub(crate) color: cgmath::Vector3<f32>,
pub(crate) tex_coord: cgmath::Vector2<f32>,

// Transformations
let identity = <cgmath::Matrix4<f32> as cgmath::SquareMatrix>::identity();

let model = <cgmath::Matrix4<f32>>::from_angle_z(cgmath::Deg(time * 10.0)) * identity;

let view = <cgmath::Matrix4<f32>>::look_at_rh(
    cgmath::Point3::new(2.0, 2.0, 2.0),
    cgmath::Point3::new(0.0, 0.0, 0.0),
    cgmath::Vector3::unit_z(),
);

let aspect_ratio =
    self.data.swapchain_extent.width as f32 / self.data.swapchain_extent.height as f32;
let mut proj = cgmath::perspective(cgmath::Deg(45.0), aspect_ratio, 0.1, 10.0);

proj.y.y *= -1.0;

// These stays the same except using cgmath
cgmath::vec3(1.0, 1.0, 1.0),

// Vec2, Vec3, Mat4 struts are replaced by Vector2, Vector3, Matrix4
```

**Comment by AuracleTech** [comment]
At the end of [Chapter 37](https://kylemayes.github.io/vulkanalia/dynamic/secondary_command_buffers.html) it says `results in sigificant performance wins`
Small typo

**Comment by AuracleTech** [comment]
[Chapter 37](https://kylemayes.github.io/vulkanalia/dynamic/secondary_command_buffers.html) has a big memory leak.
I saw [another project](https://github.com/KyleMayes/vulkanalia/blob/d276e9e1e9b0842a871097f7492c1424227e9b80/tutorial/src/32_secondary_command_buffers.rs#L46) using Vulkanalia who also discarded this line. I am unsure of it's purpose.
`self.data.secondary_command_buffers.resize_with(image_index + 1, Vec::new);`

**Comment by KyleMayes** [comment]
Thank you for noting all of this down, I am tracking it here: #151

**Comment by KyleMayes** [comment]
I've addressed most of these, but I had a question about one of the things you mentioned:

On the `vk::CommandBufferUsageFlags::SIMULTANEOUS_USE` flag from the command buffers chapter, did you also see the same errors/warnings when running the code samples that also don't use this flag? 

These command buffers are not expected to be used simultaneously as there should be one for each framebuffer. If you do see the same issue with the code samples, could you provide more info about your platform?

Thanks.

**Comment by AuracleTech** [comment]
Yes I always used the code from the tutorial except when there was errors.
This is [the commit](https://github.com/AuracleTech/titan/tree/90e85cc0ca817b0e40e611d640062b41fc21def9), you can download it, compile as is, test then remove the flag and compile again you will see the error pop up.
Hopefully this helps! I'm on Windows 10 x64 version 10.0.19045
